wgpu-core/src/command/transfer.rs

```diff
@@ -82,6 +83,158 @@ impl TextureCopyView {
+    let copy_depth = copy_size.depth as BufferAddress;
+
+    let offset = layout.offset;
+    let rows_per_image = layout.rows_per_image as BufferAddress;
+    let bytes_per_row = layout.bytes_per_row as BufferAddress;
+
+    // TODO: Once compressed textures are supported, these needs to be fixed
+    let block_width: BufferAddress = 1;
+    let block_height: BufferAddress = 1;
+    let block_size = bytes_per_texel;
+
+    let bytes_in_a_complete_row = block_size * copy_width / block_width;
+    let required_bytes_in_copy = if copy_width == 0 || copy_height == 0 || copy_depth == 0 {
+        0
+    } else {
+        let texel_block_rows_per_image = rows_per_image / block_height;
+        let bytes_per_image = bytes_per_row * texel_block_rows_per_image;
+        let bytes_in_last_slice = bytes_per_row * (copy_height / block_height - 1)
+            + (copy_width / block_width * block_size);
+        bytes_per_image * (copy_depth - 1) + bytes_in_last_slice
+    };
+
+    if rows_per_image != 0 {
+        assert!(
+            rows_per_image >= copy_height,
+            "Rows per image {} must be greater or equal to copy_extent.height {}",
+            rows_per_image,
+            copy_height
+        )
+    }
+    assert!(
+        offset + required_bytes_in_copy <= buffer_size,
+        "Texture copy using buffer indices {}..{} would overrun buffer of size {}",
+        offset,
+        offset + required_bytes_in_copy,
+        buffer_size
+    );
+    assert_eq!(
+        rows_per_image % block_height,
+        0,
+        "Rows per image {} must be a multiple of image format block height {}",
+        rows_per_image,
+        block_height,
+    );
+    assert_eq!(
+        offset % block_size,
+        0,
+        "Buffer offset {} must be a multiple of image format block size {}",
+        offset,
+        block_size,
+    );
+    if copy_height > 1 {
+        assert!(
+            bytes_per_row >= bytes_in_a_complete_row,
+            "Bytes per row {} must be at least the size of {} {}-byte texel blocks ({})",
+            bytes_per_row,
+            copy_width / block_width,
+            block_size,
+            bytes_in_a_complete_row,
+        )
+    }
+    if copy_depth > 1 {
+        assert!(
+            rows_per_image >= copy_height,
+            "Rows per image {} must be at least the desired copy height {}",
+            rows_per_image,
+            copy_height,
+        )
+    }
+}
+
+// Function copied with minor modifications from webgpu standard https://gpuweb.github.io/gpuweb/#valid-texture-copy-range
+pub(crate) fn validate_texture_copy_range(
+    texture_copy_view: &TextureCopyView,
+    texture_dimension: hal::image::Kind,
+    copy_size: &Extent3d,
+) {
+    // TODO: Once compressed textures are supported, these needs to be fixed
+    let block_width: u32 = 1;
+    let block_height: u32 = 1;
+
+    let (resource_x_size, resource_y_size, resource_z_size) = match texture_dimension {
```

Comment:
this is a bit tricky - instead of getting the texture dimensions as is, we need to consider them at the given mipmap level. There is a [level_extent](https://docs.rs/gfx-hal/0.5.0/gfx_hal/image/enum.Kind.html#method.level_extent) function in gfx-hal, but the trick is that it doesn't give you the array size. So for 2D textures, we'll need to grab the array layers explicitly:
```rust
let mut extent = texture_dimension.level_extent(texture_copy_view.mip_level);
match texture_dimension {
  hal::image::Kind::D2(_, _, array_layers, _) => extent.z = array_layers,
  hal::image::Kind::D1(..) |
  hal::image::Kind::D3(..) => (),
}
```